src/lib/transport.js

```diff
@@ -0,0 +1,290 @@
+const co = require('co')
+const debug = require('debug')('ilp:transport')
+const assert = require('assert')
+const base64url = require('../utils/base64url')
+const BigNumber = require('bignumber.js')
+const { safeConnect } = require('../utils')
+const { createDetails, parseDetails } = require('../utils/details')
+
+function _safeDecrypt (data, secret) {
+  if (!data) return {}
+  try {
+    return parseDetails({ details: data, secret })
+  } catch (err) {
+    debug('decryption error="' + err.message + '"', 'data="' + data + '"')
+    return undefined
+  }
+}
+
+function createPacketAndCondition ({
+  id,
+  destinationAmount,
+  destinationAccount,
+  secret,
+  data,
+  headers,
+  unsafeHeaders,
+  expiresAt
+}, protocol) {
+  assert(typeof destinationAmount === 'string', 'destinationAmount must be a string')
+  assert(typeof destinationAccount === 'string', 'destinationAccount must be a string')
+  assert(Buffer.isBuffer(secret), 'secret must be a buffer')
+
+  const receiverId = base64url(cryptoHelper.getReceiverId(secret))
+  const address = destinationAccount + '.~' + protocol + '.' + receiverId +
+    (id ? ('.' + id) : '')
+
+  const details = createDetails({
+    unsafeHeaders: Object.assign({}, unsafeHeaders),
+    headers: Object.assign({
+      'Expires-At': expiresAt
+    }, headers),
+
+    data: data,
+    secret: secret
+  })
+
+  const packet = Packet.serialize({
+    account: address,
+    amount: destinationAmount,
+    data: details
+  })
+
+  const condition = base64url(cc.toCondition(
+    cryptoHelper.hmacPacketForPskCondition(
+      packet,
+      secret)))
+
+  return {
+    packet,
+    condition
+  }
+}
+
+function _reject (plugin, id, reason) {
+  return plugin
+    .rejectIncomingTransfer(id, Object.assign({
+      triggered_by: plugin.getAccount(),
+      triggered_at: moment().format(),
+      additional_info: {}
+    }, reason))
+    .then(() => reason)
+}
+
+function * listen (plugin, {
+  secret,
+  allowOverPayment
+}, callback, protocol) {
+  assert(plugin && typeof plugin === 'object', 'plugin must be an object')
+  assert(typeof callback === 'function', 'callback must be a function')
+  assert(Buffer.isBuffer(secret), 'opts.secret must be a buffer')
+
+  yield safeConnect(plugin)
+
+  /**
+   * When we receive a transfer notification, check the transfer
+   * and try to fulfill the condition (which will only work if
+   * it corresponds to a request or shared secret we created)
+   * Calls the `reviewPayment` callback before fulfillingthe.
+   *
+   * Note return values are only for testing
+   */
+  function * autoFulfillCondition (transfer) {
+    // TODO: should this just be included in this function?
+    const err = yield _validateOrRejectTransfer({
+      plugin,
+      transfer,
+      protocol,
+      allowOverPayment,
+      secret
+    })
+
+    if (err) return err
+
+    const preimage = cryptoHelper.hmacPacketForPskCondition(
+      Packet.getFromTransfer(transfer),
+      secret)
```

Comment:
It doesn't seem like there's a difference in security either way. What's the reasoning for using the payment key?